src/spikeinterface/core/generate.py

```diff
@@ -220,16 +221,340 @@ def add_synchrony_to_sorting(sorting, sync_event_ratio=0.3, seed=None):
+
+            # Remove spikes that are in the refractory period.
+            violations = np.where(np.diff(injected_spike_train) < t_r)[0]
+            injected_spike_train = np.delete(injected_spike_train, violations)
+
+            # Remove spikes that violate the refractory period of the real spikes.
+            # TODO: Need a better & faster way than this.
+            min_diff = np.min(np.abs(injected_spike_train[:, None] - spike_train[None, :]), axis=1)
+            violations = min_diff < t_r
+            injected_spike_train = injected_spike_train[~violations]
+
+            if len(injected_spike_train) > n_injection:
+                injected_spike_train = np.sort(np.random.choice(injected_spike_train, n_injection, replace=False))
+
+            injected_spike_trains[segment_index][unit_id] = injected_spike_train
+
+    return NumpySorting.from_unit_dict(injected_spike_trains, sorting.get_sampling_frequency())
+
+
+class TransformSorting(BaseSorting):
+    """
+    Generates a sorting object keeping track of added spikes/units from an external spike_vector.
+    More precisely, the TransformSorting objects keeps two internal arrays added_spikes_from_existing_units and
+    added_spikes_from_new_units as boolean mask to track (in the representation as a spike vector) where
+    modifications have been made
+
+    Parameters
+    ----------
+    sorting : BaseSorting
+        The sorting object
+    added_spikes_existing_units : np.array (spike_vector)
+        The spikes that should be added to the sorting object, for existing units
+    added_spikes_new_units: np.array (spike_vector)
+        The spikes that should be added to the sorting object, for new units
+    new_units_ids: list
+        The unit_ids that should be added if spikes for new units are added
+    refractory_period_ms : float, default None
+        The refractory period violation to prevent duplicates and/or unphysiological addition
+        of spikes. Any spike times in added_spikes violating the refractory period will be
+        discarded
+
+    Returns
+    -------
+    sorting : TransformSorting
+        The sorting object with the added spikes and/or units
+    """
+
+    def __init__(
+        self,
+        sorting: BaseSorting,
+        added_spikes_existing_units=None,
+        added_spikes_new_units=None,
+        new_unit_ids: Optional[List[Union[str, int]]] = None,
+        refractory_period_ms: Optional[float] = None,
+    ):
+        sampling_frequency = sorting.get_sampling_frequency()
+        unit_ids = list(sorting.get_unit_ids())
+
+        if new_unit_ids is not None:
+            assert ~np.any(
+                np.isin(new_unit_ids, sorting.unit_ids)
+            ), "some units ids are already present. Consider using added_spikes_existing_units"
+            if len(new_unit_ids) > 0:
+                assert type(unit_ids[0]) == type(new_unit_ids[0]), "unit_ids should have the same type"
+                unit_ids += list(new_unit_ids)
```

Comment:
We need to avoid the iplace modification here
```suggestion
                unit_ids = unit_ids + list(new_unit_ids)
```